core/lib/state/src/storage_factory.rs

```diff
@@ -0,0 +1,188 @@
+            .await
+            .context("Failed to catch up state keeper RocksDB storage to Postgres")?;
+        let Some(rocksdb) = rocksdb else {
+            tracing::info!("Synchronizing RocksDB interrupted");
+            return Ok(None);
+        };
+        let rocksdb_l1_batch_number = rocksdb.l1_batch_number().await.unwrap_or_default();
+        if l1_batch_number != rocksdb_l1_batch_number {
+            anyhow::bail!(
+                "RocksDB synchronized to L1 batch #{} while #{} was expected",
+                rocksdb_l1_batch_number,
+                l1_batch_number
+            );
+        }
+        tracing::debug!(%rocksdb_l1_batch_number, "Using RocksDB-based storage");
+        Ok(Some(rocksdb.into()))
+    }
+}
+
+impl ReadStorage for PgOrRocksdbStorage<'_> {
+    fn read_value(&mut self, key: &StorageKey) -> zksync_types::StorageValue {
+        match self {
+            Self::Postgres(postgres) => postgres.read_value(key),
+            Self::Rocksdb(rocksdb) => rocksdb.read_value(key),
+            Self::RocksdbWithMemory(rocksdb, diffs, _) => {
+                match diffs.iter().rev().find_map(|x| x.get(key)) {
+                    None => rocksdb.read_value(key),
+                    Some(value) => value.value.clone(),
+                }
+            }
+        }
+    }
+
+    fn is_write_initial(&mut self, key: &StorageKey) -> bool {
+        match self {
+            Self::Postgres(postgres) => postgres.is_write_initial(key),
+            Self::Rocksdb(rocksdb) => rocksdb.is_write_initial(key),
+            Self::RocksdbWithMemory(rocksdb, diffs, _) => {
+                match diffs.iter().rev().find_map(|x| x.get(key)) {
+                    None => rocksdb.is_write_initial(key),
+                    Some(_) => false,
+                }
+            }
+        }
+    }
+
+    fn load_factory_dep(&mut self, hash: H256) -> Option<Vec<u8>> {
+        match self {
+            Self::Postgres(postgres) => postgres.load_factory_dep(hash),
+            Self::Rocksdb(rocksdb) => rocksdb.load_factory_dep(hash),
+            Self::RocksdbWithMemory(rocksdb, _, new_factory_deps) => {
+                match new_factory_deps.iter().rev().find_map(|x| x.get(&hash)) {
+                    None => rocksdb.load_factory_dep(hash),
+                    Some(value) => Some(value.clone()),
+                }
+            }
+        }
+    }
+
+    fn get_enumeration_index(&mut self, key: &StorageKey) -> Option<u64> {
+        match self {
+            Self::Postgres(postgres) => postgres.get_enumeration_index(key),
+            Self::Rocksdb(rocksdb) => rocksdb.get_enumeration_index(key),
+            Self::RocksdbWithMemory(rocksdb, diffs, _) => {
+                match rocksdb.get_enumeration_index(key) {
+                    None => diffs
+                        .iter()
+                        .find_map(|x| x.get(key).and_then(|sv| sv.enum_index)),
+                    Some(value) => Some(value),
```

Comment:
nit: I think you should firstly search for `key` in diffs and only then rocksdb

Reply:
This is the order in which RocksDB synchronization did it so I assumed there is some reasoning behind it, but I can change it for sure